packages/react-reconciler/src/ReactFiberClassComponent.js

```diff
@@ -647,6 +780,13 @@ export default function(
       newState = oldState;
     }
 
+    if (partialState != null) {
+      // Render-phase updates (like this) should not be added to the update queue,
+      // So that multiple render passes do not enqueue multiple updates.
+      // Instead, just synchronously merge the returned state into the instance.
+      newState = Object.assign({}, newState, partialState);
```

Comment:
If `newState` is null or undefined, we should avoid creating a new state object, so that it's referentially identical.

Related: if the props haven't changed, we shouldn't call this method. Same as `componentWillReceiveProps`. Need a test for this.

Reply:
> If `newState` is null or undefined, we should avoid creating a new state object, so that it's referentially identical.

Gotcha. 👍 

> Related: if the props haven't changed, we shouldn't call this method. Same as `componentWillReceiveProps`. Need a test for this.

Good observation! Fixed and added test in 361a2cf11.